File: practices/interview-bit/two-pointers/counting-triangles.py
class Solution:
    # @param A : list of integers
    # @return an integer
    def nTriang(self, A):
        n = len(A)
        A.sort(reverse = True)
        count = 0
        for i in range(0, n - 2):
            j = i + 1
            k = n - 1
            while j < k:
                sum = A[j] + A[k]
                if sum <= A[i]:
                    k -= 1
                else:
                    count += (k - j)
                    j += 1
        return count % 1000000007

File: practices/interview-bit/two-pointers/counting-triangles.cpp
// O(N^2) in time
// O(1) in space
int Solution::nTriang(vector<int> &A) {
    sort(A.rbegin(), A.rend());
    int ans = 0;
    int MOD = 1e9 + 7;
    for(int i = 0; i < (int)A.size() - 2; i++) {
        // Count A[j] + A[k] > A[i]
        int j = i + 1;
        int k = (int)A.size() - 1;
        while(j < k) {
            int sum = A[j] + A[k];
            if (sum <= A[i]) k--;
            else {
                ans = (ans + (k - j)) % MOD;
                j++;
            }
        }
    }
    return ans;
}
